rust-runtime/aws-smithy-http-server/src/request/request_id.rs

```diff
@@ -0,0 +1,214 @@
+use tower::{Layer, Service};
+use uuid::Uuid;
+
+use crate::{body::BoxBody, response::IntoResponse};
+
+use super::{internal_server_error, FromParts};
+
+/// Opaque type for Server Request IDs.
+///
+/// If it is missing, the request will be rejected with a `500 Internal Server Error` response.
+#[derive(Clone, Debug)]
+pub struct ServerRequestId {
+    id: Uuid,
+}
+
+/// The server request ID has not been added to the [`Request`](http::Request) or has been previously removed.
+#[non_exhaustive]
+#[derive(Debug, Error)]
+#[error("the `ServerRequestId` is not present in the `http::Request`")]
+pub struct MissingServerRequestId;
+
+impl ServerRequestId {
+    pub fn new() -> Self {
+        Self { id: Uuid::new_v4() }
+    }
+}
+
+impl Display for ServerRequestId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.id)
+    }
+}
+
+impl<P> FromParts<P> for ServerRequestId {
+    type Rejection = MissingServerRequestId;
+
+    fn from_parts(parts: &mut Parts) -> Result<Self, Self::Rejection> {
+        parts.extensions.remove().ok_or(MissingServerRequestId)
+    }
+}
+
+impl Default for ServerRequestId {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Clone)]
+pub struct ServerRequestIdProvider<S> {
+    inner: S,
+}
+
+pub struct ServerRequestIdProviderLayer {}
+
+impl ServerRequestIdProviderLayer {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl Default for ServerRequestIdProviderLayer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<S> Layer<S> for ServerRequestIdProviderLayer {
+    type Service = ServerRequestIdProvider<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        ServerRequestIdProvider { inner }
+    }
+}
+
+impl<R, S> Service<http::Request<R>> for ServerRequestIdProvider<S>
+where
+    S: Service<http::Request<R>>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: http::Request<R>) -> Self::Future {
+        req.extensions_mut().insert(ServerRequestId::new());
+        self.inner.call(req)
+    }
+}
+
+impl<Protocol> IntoResponse<Protocol> for MissingServerRequestId {
+    fn into_response(self) -> http::Response<BoxBody> {
+        internal_server_error()
+    }
+}
+
+/// The Client Request ID.
+///
+/// If it is missing, the request will be rejected with a `500 Internal Server Error` response.
+#[derive(Clone, Debug)]
+pub struct ClientRequestId {
+    id: Box<str>,
+}
+
+/// The client request ID has not been added to the [`Request`](http::Request) or has been previously removed.
+#[non_exhaustive]
+#[derive(Debug, Error)]
+#[error("the `ClientRequestId` is not present in the `http::Request`")]
+pub struct MissingClientRequestId;
+
+impl ClientRequestId {
+    pub fn new(id: Box<str>) -> Self {
```

Comment:
Perhaps lots of `From` implementations for `From<Box<str>>`, `From<String>` and `From<Cow<'static, str>>`.

Reply:
What about `pub fn new(id: impl Into<&str>)` or something along the lines?